botocore/serialize.py

```diff
@@ -596,9 +597,64 @@ def _convert_header_value(self, shape, value):
         else:
             return value
 
+    def _finalize_serialization(self, serialized, operation_model):
+        """Perform any required post-processing steps for serialization."""
+        self._prepare_additional_traits(serialized, operation_model)
+
 
 class RestJSONSerializer(BaseRestSerializer, JSONSerializer):
 
+    def _finalize_serialization(self, serialized, operation_model):
+        super()._finalize_serialization(serialized, operation_model)
+
+        self._set_default_json_body(serialized, operation_model)
+        self._set_default_content_type_header(serialized, operation_model)
```

Comment:
I think the current implementation is correct, but I'm concerned that we're going about this from the wrong side. I'm not sure it makes sense to set a "default" retroactively like this. It feels like we're building up this list of predicates based off of the serializer's current behavior and finding the hole where we should put the default JSON body. This feels like a hard pattern to maintain because now you have to update the serializer's logic, but also make sure the assumptions being made in this finalization step still hold.

Can we not handle this more proactively before / during serialization? My hunch is that we should be able to set this payload/content-type up front and then have the serializer override the body/content-type if it knows better. This seems preferential to having the serializer implement some logic `A` around the model and then have this post serialization step where we check for `not A`.

Reply:
>  It feels like we're building up this list of predicates based off of the serializer's current behavior and finding the hole where we should put the default JSON body. This feels like a hard pattern to maintain because now you have to update the serializer's logic, but also make sure the assumptions being made in this finalization step still hold.

Completely agree, I'm not a fan of this pattern, but it's unfortunately a necessity with our current inheritance model. We've made it difficult to modify rest-json specific behavior without moving almost all of `BaseRestSerializer` down into this class. I'll see if we can pull a bit more of this up by adding some inline helper functions to `serialize_to_request` and the other methods.